text_extensions_for_pandas/io/conll.py

```diff
@@ -217,6 +301,122 @@ def _parse_conll_file(input_file: str,
+    :param column_names: Names for the metadata columns that come after the
+     token text. These names will be used to generate the names of the dataframe
+     that this function returns.
+    :param iob_columns: Mask indicating which of the metadata columns after the
+     token text should be treated as being in IOB format. If a column is in IOB format,
+     the returned data structure will contain *two* columns, holding IOB tags and
+     entity type tags, respectively. For example, an input column "ent" will turn into
+     output columns "ent_iob" and "ent_type".
+    :param predicate_args: whether or not predicate arguments are stored in this file format.
+
+    :returns: A list of lists of _SentenceData objects. The top list has one entry per
+     document. The next level lists have one entry per sentence.
+    """
+    with open(input_file, "r") as f:
+        lines = f.readlines()
+
+    # Build up a list of document metadata as Python objects
+    docs = []  # Type: List[List[Dict[str, List[str]]]]
+
+    current_sentence = _SentenceData(column_names.copy(), iob_columns.copy(),predicate_args)
+
+    # Information about the current document
+    sentences = []  # Type: SentenceData
+
+    # metadata specific to conll_u
+    doc_id = ''
+    paragraph_id = ''
+
+    # if we merge subtokens we need additional logic
+    in_subtok = False  # set this flag when inside of subtoken
+    subtok_end = None  # only valid when in subtok
+
+    for i in range(len(lines)):
+        line = lines[i].strip()
+        if 0 == len(line):
+            # Blank line is the sentence separator
+            if current_sentence.num_tokens > 0:
+                sentences.append(current_sentence)
+                current_sentence = _SentenceData(column_names.copy(), iob_columns.copy(),predicate_args)
+                current_sentence.set_conll_u_metadata(doc_id=doc_id,
+                                                      paragraph_id=paragraph_id)  # sentence id gets set later
+        elif line[0] == '#':
+            # TODO: this is metadata. We might want to store it.
+            line_elems = line.split(' = ')
+            if line_elems[0] == _EWT_DOC_SEPERATOR:
+                if i > 0:
+                    # End of document.  Wrap up this document and start a new one.
+                    #
+                    docs.append(sentences)
+                    sentences = []
+                    # reset doc, paragraph and sentence id's
+                doc_id = line_elems[1]
+                current_sentence.set_conll_u_metadata(doc_id=doc_id)
+                paragraph_id = ''
+            elif line_elems[0] == "# newpar id":
+                paragraph_id = line_elems[1]
+                current_sentence.set_conll_u_metadata(paragraph_id=paragraph_id)
+            elif line_elems[0] == "# sent_id":
+                sentence_id = line_elems[1]
+                current_sentence.set_conll_u_metadata(sent_id=sentence_id)
+
```

Comment:
There should be an additional branch to this `if` statement that gathers up any other metadata key-value pairs that the dataset sees fit to attach to the sentence (see https://universaldependencies.org/format.html#sentence-boundaries-and-comments).

It would be nice if there was also a way to have this function map a user-configurable set of additional metadata values to additional columns of the returned DataFrame.